### PR TITLE
[luci/service] Support shape inference for non-const paddings

### PR DESCRIPTION
**Pull request by icodo98**
This PR supports infer shape of pad operation with dynamic paddings.
ONE-DCO-1.0-Signed-off-by: JuYoung Lee rsb98759@gmail.com

### CODE REVIEW

**Comment by icodo98** [comment]
related to : #13790 
draft : #13877

File: compiler/luci/service/src/CircleShapeInferenceHelper.cpp [hunk 1]
```diff
@@ -180,6 +180,12 @@ loco::TensorShape pad_shape(const loco::TensorShape &input_shape, const luci::Ci
   loco::TensorShape output_shape;
 
   output_shape.rank(input_shape.rank());
+
+  auto const_padding = dynamic_cast<const luci::CircleConst *>(paddings);
+
+  if (const_padding == nullptr)
+    return output_shape;
```
Comment (seanshpark) [inline compiler/luci/service/src/CircleShapeInferenceHelper.cpp:187]:
```suggestion
  auto const_padding = dynamic_cast<const luci::CircleConst *>(paddings);
  if (const_padding == nullptr)
    return output_shape;
```

File: compiler/luci/service/src/CircleShapeInferenceHelper.h [hunk 2]
```diff
@@ -49,8 +49,8 @@ loco::TensorShape circle_shape(const luci::CircleNode *node);
 loco::TensorShape broadcast_shape(const loco::TensorShape &x, const loco::TensorShape &y);
 
 // Return shape of pad ops using paddings.
-loco::TensorShape pad_shape(const loco::TensorShape &input_shape,
-                            const luci::CircleConst *paddings);
+// If paddings is static, return the shape filled with unknown dimensions.
```
Comment (seanshpark) [inline compiler/luci/service/src/CircleShapeInferenceHelper.h:52]: Q) what if paddings is not static? does it return shape filled with known dimensions?

File: compiler/luci/service/src/Nodes/CirclePad.cpp [hunk 3]
```diff
@@ -32,8 +32,9 @@ namespace sinf
 
 loco::TensorShape Algorithm::visit(const luci::CirclePad *node)
 {
-  // TODO support non-const case
-  auto paddings = loco::must_cast<luci::CircleConst *>(node->paddings());
+
+  auto paddings = loco::must_cast<const luci::CircleNode *>(node->paddings());
+
```
Comment (seanshpark) [inline compiler/luci/service/src/Nodes/CirclePad.cpp:37]:
```suggestion
  auto paddings = loco::must_cast<const luci::CircleNode *>(node->paddings());
```

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp [hunk 4]
```diff
@@ -90,3 +90,60 @@ TEST(ShapeRuleTest, pad_without_padding_NEG)
...
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  node_paddings->dtype(loco::DataType::S64);
+  node_paddings->shape({4, 2});
+  node_paddings->shape_status(luci::ShapeStatus::VALID);
+
+  const loco::DataType S64 = loco::DataType::S64;
+  uint32_t t = 64 * 8;
```
Comment (seanshpark) [inline compiler/luci/service/src/Nodes/CirclePad.test.cpp:143]: Q) what is the meaning of `64 * 8`?

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp [hunk 5]
```diff
@@ -90,3 +90,60 @@ TEST(ShapeRuleTest, pad_without_padding_NEG)
...
+  ASSERT_FALSE(shape.dim(2).known());
+  ASSERT_FALSE(shape.dim(3).known());
+
+  ASSERT_EQ(0, shape.dim(0).value());
+  ASSERT_EQ(0, shape.dim(1).value());
+  ASSERT_EQ(0, shape.dim(2).value());
+  ASSERT_EQ(0, shape.dim(3).value());
+}
+
+TEST(ShapeRuleTest, pad_without_input_NEG)
```
Comment (seanshpark) [inline compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: I can't catch what false case of the modified code is for this test case.
Can you please point it out?

File: compiler/luci/service/src/CircleShapeInferenceHelper.h (hunk 2 above)
Comment (icodo98) [reply compiler/luci/service/src/CircleShapeInferenceHelper.h:52]: Oh, It's typo... 😢
It return shape filled with unknown dimensions if paddings is **not** static

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp (hunk 5 above)
Comment (icodo98) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: Pad operation has two args, input and paddings. 
In this case, I intentionally excluding input.

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp (hunk 4 above)
Comment (icodo98) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:143]: This line is not necessary.
It's just ensure that `padddings` size is large enough to be used.

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp (hunk 5 above)
Comment (seanshpark) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: I mean, from your changes. What change is related with this negative test?

File: compiler/luci/service/src/Nodes/CirclePad.cpp [hunk 6]
```diff
@@ -32,10 +32,10 @@ namespace sinf
 
 loco::TensorShape Algorithm::visit(const luci::CirclePad *node)
 {
-  // TODO support non-const case
-  auto paddings = loco::must_cast<luci::CircleConst *>(node->paddings());
+  auto paddings = loco::must_cast<const luci::CircleNode *>(node->paddings());
   auto circle_input = loco::must_cast<const luci::CircleNode *>(node->input());
   auto input_shape = circle_shape(circle_input);
+
```
Comment (seanshpark) [inline compiler/luci/service/src/Nodes/CirclePad.cpp:38]:
```suggestion
```

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp (hunk 5 above)
Comment (icodo98) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: > I mean, from your changes. What change is related with this negative test?

This PR infers the output shape through the shape of the input when `paddings` is not static, so inference is impossible if there is no input. 
So I created a case with normal padding but no input.

If you feel that this TC is not clear, let us consider another TC.

Comment (seanshpark) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: My first comment was `Can you please point it out?`

Comment (icodo98) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: > My first comment was `Can you please point it out?` The code you introduced, which one is it?

```
 node_pad->paddings(node_paddings);

 ASSERT_ANY_THROW(shape_inf_rule.infer(node_pad, shape))
```
only paddings is set for `CirclePad` so the input is `nullptr`.

Adding `node_pad->input(nullptr)` line might better clarify my intention.

Comment (seanshpark) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: Not the test code, the code that you changed in shape inference.

Comment (icodo98) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: The code I changed in shape inference added new support for non-static `paddings` as follows.

```
  output_shape.rank(input_shape.rank());

  auto const_padding = dynamic_cast<const luci::CircleConst *>(paddings);
  if (const_padding == nullptr)
    return output_shape;
```

In this code, `input_shape.rank()` is used to determine the rank of the output, so input is required.

Comment (seanshpark) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: image

`output_shape.rank(input_shape.rank());` was existing code, and is not changed,
and this line is related but doesn't seem to be "directly" related with `input should exist`.
It's hard to understand with indirect relation.

Comment (icodo98) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: > `output_shape.rank(input_shape.rank());` was existing code, and is not changed, and this line is related but doesn't seem to be "directly" related with `input should exist`. It's hard to understand with indirect relation.

As you mentioned, the current TC seems to have a weak connection with the code I modified.

To fix this, how about changing TC that instead `input`, change `paddings` to `nullptr`?
   -> In this case, it would serve as a false case where `paddings` must be a `circleNode` at the very least.

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp [hunk 7]
```diff
@@ -90,3 +90,56 @@ TEST(ShapeRuleTest, pad_without_padding_NEG)
 
   ASSERT_ANY_THROW(shape_inf_rule.infer(node_pad, shape));
 }
+
+TEST(ShapeRuleTest, pad_non_const_paddings)
```
Comment (shs-park) [inline compiler/luci/service/src/Nodes/CirclePad.test.cpp:94]: @icodo98, @seanshpark,

IMHO, it seems to me that this PR implements handling for the exception case for Non-const `padding` node.

So the case related to this implementation can be considered as a `NEG` test,
so this itself can be considered as a `NEG` test already.
What do you think?

```suggestion
TEST(ShapeRuleTest, pad_non_const_paddings_NEG)
```

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp [hunk 8]
```diff
@@ -90,3 +90,56 @@ TEST(ShapeRuleTest, pad_without_padding_NEG)
...
+  ASSERT_EQ(0, shape.dim(2).value());
+  ASSERT_EQ(0, shape.dim(3).value());
+}
+
+TEST(ShapeRuleTest, pad_empty_padding_NEG)
+{
+  auto g = loco::make_graph();
+  auto node_pad = g->nodes()->create<luci::CirclePad>();
+  auto node_input = g->nodes()->create<luci::CircleInput>();
+  auto node_paddings = g->nodes()->create<luci::CircleAdd>();
```
Comment (shs-park) [inline compiler/luci/service/src/Nodes/CirclePad.test.cpp:133]: How about adding a comment to the intended test?
```suggestion
  // the padding node is another "non-const" case (i.e., Add)
  auto node_paddings = g->nodes()->create<luci::CircleAdd>();
```

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp [hunk 9]
```diff
@@ -90,3 +90,56 @@ TEST(ShapeRuleTest, pad_without_padding_NEG)
...
+  ASSERT_FALSE(shape.dim(2).known());
+  ASSERT_FALSE(shape.dim(3).known());
+
+  ASSERT_EQ(0, shape.dim(0).value());
+  ASSERT_EQ(0, shape.dim(1).value());
+  ASSERT_EQ(0, shape.dim(2).value());
+  ASSERT_EQ(0, shape.dim(3).value());
+}
+
+TEST(ShapeRuleTest, pad_empty_padding_NEG)
```
Comment (shs-park) [inline compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: You need to change the name of this test.

```suggestion
TEST(ShapeRuleTest, pad_non_const_node_padding_NEG)
```

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp [hunk 10]
```diff
@@ -90,3 +90,56 @@ TEST(ShapeRuleTest, pad_without_padding_NEG)
...
+  luci::sinf::Rule shape_inf_rule;
+
+  node_input->shape({1, 2, 3, 4});
+  node_input->shape_status(luci::ShapeStatus::VALID);
+
+  node_pad->input(node_input);
+  node_pad->paddings(node_paddings);
+
+  ASSERT_FALSE(shape_inf_rule.infer(node_pad, shape));
+}
```
Comment (shs-park) [inline compiler/luci/service/src/Nodes/CirclePad.test.cpp:145]: You may remove this test case if above `pad_non_const_paddings` case could be considered a negative test.

**Review by shs-park — COMMENTED** [review]
I left some questions and comments.
PTAL.
=)

File: compiler/luci/service/src/CircleShapeInferenceHelper.cpp [hunk 11]
```diff
@@ -180,6 +180,11 @@ loco::TensorShape pad_shape(const loco::TensorShape &input_shape, const luci::Ci
   loco::TensorShape output_shape;
 
   output_shape.rank(input_shape.rank());
+
+  auto const_padding = dynamic_cast<const luci::CircleConst *>(paddings);
+  if (const_padding == nullptr)
+    return output_shape;
```
[the remark this answers is not in the document]
Comment (icodo98) [reply compiler/luci/service/src/CircleShapeInferenceHelper.cpp:186]: > Just to clarify, does this mean that if the padding node is not constant, it does nothing, right?

Yes. when the `paddings` is not constant, it does nothing and return output shape filled with unknown dimensions.

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp (hunk 7 above)
Comment (icodo98) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:94]: > IMHO, it seems to me that this PR implements handling for the exception case for Non-const `padding` node.

I completely agree with that opinion. 😊

Comment (seanshpark) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:94]: Not sure, from our SE, definition of `negative` was ambiguous to me.

From the Pad Op,
- padding can be constant
- padding can be non constant

I think these two are normal cases.
For the point of return shape, now shape can have unknown dims as normal cases.
So, as we treat non constant and return shape with all unknown dims looks normal.
Strictly speaking, I don't think this is negative test.

File: compiler/luci/service/src/CircleShapeInferenceHelper.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/service/src/CircleShapeInferenceHelper.cpp:186]: Just above this line, there is 
```
output_shape.rank(input_shape.rank());
```
which implicitly sets all dimes to unknown.
In another PR, add a comment about this would help others reading the code.

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp (hunk 9 above)
Comment (icodo98) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: @seanshpark 

What do you think about this new `NEG` test?

Comment (seanshpark) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: Please discuss with @shs-park and @zetwhite :)

Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: copy-paste from https://github.com/Samsung/ONE/pull/13875#discussion_r1751156686 

I had talk with @seanshpark and @shs-park. 
We have reached a consensus not to add negative tests for `LUCI_ASSERT`.

Because... 
  * For consistency, `LUCI_ASSERT` in another files are not checked in negative test. 
  * Since `assert` works differently in `release` and `debug` build, it is a bit troublesome to write the code for `assert`. 

In short, you could remove `assert` related NEG tests. 

-----------------------------------------------

BTW, Since we should have more negative tests than positive tests (by SE policy), we need to find another place(maybe `luci/imporeter`..?) to add Neg tests. Adding another NEG tests can be work in another PRs, not in this PR. -> Let's talk about this, by making new issue.

Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:128]: @shs-park, please let me know if I misunderstood things or if i missed sth. :)

**Review by shs-park — APPROVED** [review]
LGTM!
=)

Let's create another issue to discuss additional NEG tests.

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

**Review by zetwhite — APPROVED** [review]
LGTM
